### DISCUSSION TITLE
Is there an Image View plugin for far2l?

### DISCUSSION THREAD

**Discussion by faerot** [elfmz/far2l#2389] [General]
On of the big things I miss from windows far is the ability to quickly browse images. Is there a plugin for it?

**Comment by unxed** [comment]
Very experimental attempt
https://github.com/unxed/far2l/tree/imgview

See also: https://github.com/elfmz/far2l/issues/356

**Comment by unxed** [comment]
Also far2l support images preview using chafa (F3 -> F5).
Also take a look at https://github.com/jstkdng/ueberzugpp

**Comment by faerot** [comment]
chafa is nice for ssh to servers and stuff, but I am talking about proper graphical image viewer. in windows far such plugin usually displays overlayed graphical window on top of far. one may argue that it's not really necessary if you can just press "enter" and launch graphical program, but for me this method has several drawbacks:
1) external programs cannot view images where far can step into (archives, remote servers)
2) external programs are slow to launch, even lightweight ones, especially when you are quest to quickly navigate through stuff using Far.

**Reply by unxed** [reply]
That's exactly that [ueberzugpp](https://github.com/jstkdng/ueberzugpp) does.

**Reply by faerot** [reply]
> That's exactly that [ueberzugpp](https://github.com/jstkdng/ueberzugpp) does.

Yeah. All we need is a plugin which integrates it into Far, right? :)

**Reply by unxed** [reply]
Probably

**Comment by faerot** [comment]
Displaying image is not a problem, I have created a plugin which displays an image in a window in 5 minutes (using python of course). The main chunk of work is properly integrating it's behavior into far so it will feel like native in viewer, i.e. pressing F3 opens it, pressing PgDn (or whatever button is used to go to next image) will move Far cursor on appropriate file, so when you quit the viewer, you are selecting the file you were actually viewing (which is not the case in external viewers and extremely annoying).

**Comment by faerot** [comment]
The plugin also needs to preload next and previous image(s) in advance, so when you navigate to next/previous image, it shows instantly.

**Comment by faerot** [comment]
I have played around with those options are there are some problems. Does not look like ueverzugpp provide necessary options for proper image viewer - in order to do usable image viewer, it needs to allow display image using cropping and scaling to implement zoom in, scroll, etc.

I have also played with kitty a bit more and which remote control @ is very slow, looks like ``kitten icat`` provides good speed. The problem is it does not work when invoked from far2l for some reason:
```
$ kitten icat ~/images/2560x1600.jpg
Error: Terminal does not support reporting screen sizes in pixels, use a terminal such as kitty, WezTerm, Konsole, etc. that does.
```
I run far2l in kitty and invoke kitten icat from inside far2l.
Any idea what that error is all about?

**Reply by unxed** [reply]
far2l's internal virtual terminal emulator is not the same terminal far2l is run from. far2l's terminal does not support kitty image display protocol for now.

**Comment by elfmz** [comment]
done
